### PR TITLE
Clustergram customization

### PR DESCRIPTION
**Pull request by BioEndeavor**

### CODE REVIEW

**Review by quantumsplicer — APPROVED** [review]

File: src/cats/functions/drawCatTitles.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    });
+  }
+
+  const vertArg = `
+      precision mediump float;
+      attribute vec2 position;
+
+      uniform vec2 inst_offset;
+      uniform float y_offset;
+      uniform float x_offset;
```
Comment (abiswas97) [inline src/cats/functions/drawCatTitles.js:62]: Can we please put this in a common file and import if the webgl shader code is just being copy pasted?

File: src/cats/functions/drawCatTitles.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  
+        }
+      }`;
+  const fragArg = `
+      precision mediump float;
+      void main () {
+        gl_FragColor = vec4(0.2, 0.2, 0.2, 1.0);
+      }`;
+  regl({
+    vert: vertArg,
```
Comment (abiswas97) [inline src/cats/functions/drawCatTitles.js]: spacing

File: src/draws/drawAxisComponents.js [hunk 3]
```diff
@@ -38,10 +38,7 @@ export default (function drawAxisComponents(
       });
     });
     // only show the dendrogram if the current axis is in clust ordering
-    if (
-      store.select("order").inst[inst_axis] === "clust" &&
-      store.select("order").new[inst_axis] === "clust"
-    ) {
+    if (store.select("order").new[inst_axis] === "clust") {
```
Comment (abiswas97) [inline src/draws/drawAxisComponents.js:41]: What is inst, and why did we remove the condition?

File: src/index.ts [hunk 4]
```diff
@@ -151,12 +151,13 @@ function clustergrammer_gl(
         },
         reorder: (
           axis: "row" | "col" | string,
-          order: "clust" | "sum" | "var" | "ini" | string
+          order: "clust" | "sum" | "var" | "ini" | "anno" | string
         ) => {
```
Comment (abiswas97) [inline src/index.ts]: This type is copy pasted in the next file - make it a common type

File: src/matrixLabels/makeRowTextArgs.js [hunk 5]
```diff
@@ -104,5 +105,10 @@ export default function make_row_text_args(regl, store, zoom_function) {
       range: [0, 1],
     },
   };
+
+  const rows = store.select("cat_data").row;
+  rows?.map((row, index) => {
+    drawCatTitles(regl, store, "row", row.cat_title, index);
+  });
   return args;
```
Comment (abiswas97) [inline src/matrixLabels/makeRowTextArgs.js]: Is this the code that renders the row/col labels? Did clustergrammer not have any current way to do it?

File: src/matrixLabels/vectorizeLabel.js [hunk 6]
```diff
@@ -10,6 +10,9 @@ export default function vectorize_label(
     textAlign: "left",
     triangles: true,
     size: fontDetails,
+    styletags: {
+      bolds: true, // to show any bold text
+    },
```
Comment (abiswas97) [inline src/matrixLabels/vectorizeLabel.js:15]: Why plural?

File: src/reorders/reorderCatArgs.ts [hunk 7]
```diff
@@ -17,6 +17,8 @@ export default (function reorderCatArgs(
       catArgsManager.makeNewCatArrs(store, inst_axis, cat_index);
       // update the attribute
       catArgsManager.updateCatArgsAttribute(inst_axis, cat_index);
+
+      catArgsManager.regenerateCatArgsArrs(store);
     }
```
Comment (abiswas97) [inline src/reorders/reorderCatArgs.ts]: What does this do?

File: src/ui/functions/controlPanel/buildControlPanel.js [hunk 8]
```diff
@@ -157,7 +158,7 @@ export default function build_control_panel(
       const y_trans = y_offset_buttons - 2 * button_dim.buffer + 2;
       return "translate( " + x_offset + ", " + y_trans + ")";
     });
-  const order_options = ["clust", "sum", "var", "ini"];
+  // const order_options = ["clust", "sum", "var", "ini"];
   control_svg
```
Comment (abiswas97) [inline src/ui/functions/controlPanel/buildControlPanel.js]: Remove unused code please

File: src/ui/functions/controlPanel/buildControlPanel.js [hunk 9]
```diff
@@ -270,8 +271,8 @@ export default function build_control_panel(
         "translate(" + button_dim.width / 2 + ", " + button_dim.height / 2 + ")"
       );
   });
-  buildReorderCatTitles(regl, store, catArgsManager, camerasManager);
-  buildReclusterSection(regl, store, catArgsManager, camerasManager);
+  // buildReorderCatTitles(regl, store, catArgsManager, camerasManager);
+  // buildReclusterSection(regl, store, catArgsManager, camerasManager);
   // row search
```
Comment (abiswas97) [inline src/ui/functions/controlPanel/buildControlPanel.js]: Here as well. Why were these removed?

File: src/ui/functions/controlPanel/buildReclusterSection.js [hunk 10]
```diff
@@ -181,11 +181,16 @@ export default (function build_recluster_section(
...
+    })
     .on("click", function (d) {
+      const full = this.getAttribute("data-full");
+
       dispatch(
         store.actions.mutateMatrixState({
           potential_recluster: {
-            distance_metric: d.full,
+            distance_metric: full,
           },
```
Comment (abiswas97) [inline src/ui/functions/controlPanel/buildReclusterSection.js]: What does this change help with?

File: src/ui/functions/dendrograms/buildDendrogramSliders.js [hunk 11]
```diff
@@ -29,7 +29,7 @@ export default function build_dendrogram_sliders(regl, store) {
         }
       });
     if (inst_axis === "row") {
-      axis_slider_container.style("right", "-10px").style("top", "80px");
+      axis_slider_container.style("right", "-40px").style("top", "80px");
     } else {
```
Comment (abiswas97) [inline src/ui/functions/dendrograms/buildDendrogramSliders.js:33]: We've used 40px twice = is this a common offset value?

**Review by abiswas97 — CHANGES_REQUESTED** [review]
Add comments explaining changes, not sure what changes I'm looking at.

File: src/draws/drawAxisComponents.js (hunk 3 above)
Comment (BioEndeavor) [reply src/draws/drawAxisComponents.js:41]: inst is the previous state of the button clicked and new is the current state of the button clicked, And dendrogram are added only when we select "clust" button. so this was removed so that it will only check if the new state or current state is clust or not.

File: src/matrixLabels/vectorizeLabel.js (hunk 6 above)
Comment (BioEndeavor) [reply src/matrixLabels/vectorizeLabel.js:15]: the vectorize-text takes this as the argument for showing bold, its how they have written https://github.com/mikolalysenko/vectorize-text

File: src/ui/functions/dendrograms/buildDendrogramSliders.js (hunk 11 above)
Comment (BioEndeavor) [reply src/ui/functions/dendrograms/buildDendrogramSliders.js:33]: No its not a common offset value

File: src/reorders/reorderCatArgs.ts (hunk 7 above)
Comment (BioEndeavor) [reply src/reorders/reorderCatArgs.ts]: Zooming in and out  of the matrix removed the current state of the categories and it was setting it back to the initial, so have added this to prevent that.

File: src/ui/functions/controlPanel/buildReclusterSection.js (hunk 10 above)
Comment (BioEndeavor) [reply src/ui/functions/controlPanel/buildReclusterSection.js]: This was added so that the Recluster block can work properly as the d.full was returning undefined and it was not working.

File: src/ui/functions/controlPanel/buildControlPanel.js (hunk 9 above)
Comment (BioEndeavor) [reply src/ui/functions/controlPanel/buildControlPanel.js]: buildReorderCatTitles were used to create the categorical titles so I removed it and buildReclusterSection was not used currently, only the Reorder block is used

File: src/cats/functions/drawCatTitles.js (hunk 1 above)
Comment (BioEndeavor) [reply src/cats/functions/drawCatTitles.js:62]: these are slightly different from the main webgl shader code

File: src/ui/functions/controlPanel/buildReclusterSection.js (hunk 10 above)
Comment (abiswas97) [reply src/ui/functions/controlPanel/buildReclusterSection.js]: Doesn't that break the logic though? Is d itself undefined? Not a fan of mixing DOM accessors becuase the value passed on click is wrong. Have we checked if there's an issue with the value being passed on click?

File: src/ui/functions/controlPanel/buildControlPanel.js (hunk 9 above)
Comment (abiswas97) [reply src/ui/functions/controlPanel/buildControlPanel.js]: If it was being used to show the category titles, why was it removed?

File: src/cats/functions/drawCatTitles.js (hunk 1 above)
Comment (abiswas97) [reply src/cats/functions/drawCatTitles.js:62]: What is different? I was under the impression this wasn't changed

File: src/draws/drawAxisComponents.js (hunk 3 above)
Comment (abiswas97) [reply src/draws/drawAxisComponents.js:41]: If this line was functional in the original codebase, the issue would more likely be in our implementation and not here.

File: src/index.ts (hunk 4 above)
Comment (BioEndeavor) [reply src/index.ts]: did that

File: src/matrixLabels/makeRowTextArgs.js (hunk 5 above)
Comment (BioEndeavor) [reply src/matrixLabels/makeRowTextArgs.js]: Yes, clustergrammer did that outside canvas

File: src/ui/functions/controlPanel/buildControlPanel.js (hunk 8 above)
Comment (BioEndeavor) [reply src/ui/functions/controlPanel/buildControlPanel.js]: removed

File: src/reorders/reorderCatArgs.ts [hunk 12]
```diff
@@ -17,6 +17,8 @@ export default (function reorderCatArgs(
       catArgsManager.makeNewCatArrs(store, inst_axis, cat_index);
       // update the attribute
       catArgsManager.updateCatArgsAttribute(inst_axis, cat_index);
+
+      catArgsManager.regenerateCatArgsArrs(store); // added to preserve the current state fixed
     }
```
Comment (abiswas97) [inline src/reorders/reorderCatArgs.ts:22]: For better readability, can we stick to comments on the top? The line above this has the comment above it, and the comment here is to the side. Makes it hard to read.

**Review by abiswas97 — APPROVED** [review]
